packages/grafana-prometheus/src/querybuilder/components/PromQueryCodeEditorAutocompleteInfo.tsx

```diff
@@ -0,0 +1,67 @@
+import React, { useState, useEffect } from 'react';
+
+import { selectors } from '@grafana/e2e-selectors';
+import { config } from '@grafana/runtime';
+import { IconButton, Text, Stack } from '@grafana/ui';
+
+import {
+  CODE_MODE_SUGGESTIONS_INCOMPLETE_EVENT,
+  isSuggestionsIncompleteEvent,
+} from '../../components/monaco-query-field/monaco-completion-provider/data-provider';
+import { PromQueryEditorProps } from '../../components/types';
+import { QueryEditorMode } from '../shared/types';
+
+interface Props {
+  datasourceUid: PromQueryEditorProps['datasource']['uid'];
+  editorMode: QueryEditorMode;
+}
+
+export function PromQueryCodeEditorAutocompleteInfo(props: Readonly<Props>) {
+  const [autocompleteLimit, setAutocompleteLimit] = useState('n');
+  const [autocompleteLimitExceeded, setAutocompleteLimitExceeded] = useState(false);
+
+  useEffect(() => {
+    const handleSuggestionsIncompleteEvent = (e: Event) => {
+      if (!isSuggestionsIncompleteEvent(e)) {
+        return;
+      }
+
+      if (e.detail.datasourceUid === props.datasourceUid) {
+        setAutocompleteLimitExceeded(true);
+        setAutocompleteLimit(e.detail.limit.toString());
+      }
+    };
+
+    addEventListener(CODE_MODE_SUGGESTIONS_INCOMPLETE_EVENT, handleSuggestionsIncompleteEvent);
```

Comment:
I know the datasourceUid is not gonna change and if it changes, we will be in a new datasource. But is that a safe practice to have the event handler in `useEffect`? The feature is working well and not saying we should change this but just want to question if that's safe :)

Reply:
Probably best to have it in a `useCallback`, since it's relying on a prop.

Updated in 065f2bd097c3db55a09a8ede82088ffd0a2973d3 :)